### PR TITLE
Support query_named_block_nodes during live commit

### PR DESCRIPTION
**Pull request by aliang123**
blockdev_commit_query_named_block_nodes: add support
for query_named_block_nodes during live commit

Signed-off-by: Aihua Liang <aliang@redhat.com>

id: 1820429

### CODE REVIEW

**Comment by aliang123** [comment]
@ngu-niny, @zhencliu, please help to review, thanks

**Comment by aliang123** [comment]
(1/1) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.blockdev_commit_query_named_block_nodes.q35: PASS (72.70 s)

File: qemu/tests/cfg/blockdev_commit_query_named_block_nodes.cfg [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    image_name_sn3 = sn3
+    image_size_sn3 = 500M
+    image_format_sn3 = qcow2
+
+    image_name_sn4 = sn4
+    image_size_sn4 = 500M
+    image_format_sn4 = qcow2
+
+    device_tag = "data"
+    format = qcow2
```
Comment (zhencliu) [inline qemu/tests/cfg/blockdev_commit_query_named_block_nodes.cfg:34]: remote format here?

File: qemu/tests/cfg/blockdev_commit_query_named_block_nodes.cfg [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    rebase_mode = unsafe
+    query_named_block_nodes = yes
```
Comment (zhencliu) [inline qemu/tests/cfg/blockdev_commit_query_named_block_nodes.cfg:36]: This tc is designed for query blocks when committing, so remove query_named_block_nodes?

File: qemu/tests/cfg/blockdev_commit_query_named_block_nodes.cfg [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    force_create_image_data = yes
+    force_remove_image_data = yes
+    start_vm = yes
+    kill_vm = yes
+    storage_pools = default
+    storage_type_default = "directory"
+    storage_pool = default
+    image_size_data = 500M
+    image_name_data = data
+    snapshot_tags = sn1 sn2 sn3 sn4
```
Comment (zhencliu) [inline qemu/tests/cfg/blockdev_commit_query_named_block_nodes.cfg:15]: what about snapshot_tags_data? of course, snapshot_tags works well, but snapshot_tags_data explicitly shows we make snapshots for data image

File: qemu/tests/blockdev_commit_query_named_block_nodes.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        options = ["base-node", "top-node", "speed"]
+        arguments = self.params.copy_from_keys(options)
+        arguments["base-node"] = self.get_node_name(device)
+        arguments["top-node"] = self.get_node_name(snapshot_tags[-2])
+        device = self.get_node_name(snapshot_tags[-1])
+        commit_cmd = backup_utils.block_commit_qmp_cmd
+        cmd, args = commit_cmd(device, **arguments)
+        self.main_vm.monitor.cmd(cmd, args)
+        job_id = args.get("job-id", device)
+        if self.params.get("query_named_block_nodes", "no"):
```
Comment (zhencliu) [inline qemu/tests/blockdev_commit_query_named_block_nodes.py:23]: we do the query block nodes test in the case, so we may not need a flag here

File: qemu/tests/blockdev_commit_query_named_block_nodes.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            self.main_vm.monitor.cmd("query-named-block-nodes")
```
Comment (zhencliu) [inline qemu/tests/blockdev_commit_query_named_block_nodes.py:24]: Should we check the output of the query command?

File: qemu/tests/blockdev_commit_query_named_block_nodes.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        job_utils.wait_until_block_job_completed(self.main_vm, job_id)
```
Comment (zhencliu) [inline qemu/tests/blockdev_commit_query_named_block_nodes.py:25]: After commit, there is one step 'info block' in the manual case, but I didn't see it, so we can remove this step from case?

File: qemu/tests/blockdev_commit_query_named_block_nodes.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+
+def run(test, params, env):
+    """
+    Block commit base Test
+
+    1. boot guest with data disk
+    2. create 4 snapshots and save file in each snapshot
+    3. commit snapshot 3 to snapshot 4
```
Comment (ngu-niny) [inline qemu/tests/blockdev_commit_query_named_block_nodes.py:34]: Should be 'commit snapshot4 to snapshot3' instead.

File: qemu/tests/cfg/blockdev_commit_query_named_block_nodes.cfg (hunk 1 above)
Comment (aliang123) [reply qemu/tests/cfg/blockdev_commit_query_named_block_nodes.cfg]: Have updated.

File: qemu/tests/cfg/blockdev_commit_query_named_block_nodes.cfg (hunk 2 above)
Comment (aliang123) [reply qemu/tests/cfg/blockdev_commit_query_named_block_nodes.cfg]: Have update.

File: qemu/tests/cfg/blockdev_commit_query_named_block_nodes.cfg (hunk 3 above)
Comment (aliang123) [reply qemu/tests/cfg/blockdev_commit_query_named_block_nodes.cfg]: Have updated.

File: qemu/tests/blockdev_commit_query_named_block_nodes.py (hunk 4 above)
Comment (aliang123) [reply qemu/tests/blockdev_commit_query_named_block_nodes.py]: Have updated.

File: qemu/tests/blockdev_commit_query_named_block_nodes.py (hunk 5 above)
Comment (aliang123) [reply qemu/tests/blockdev_commit_query_named_block_nodes.py]: No need.
 In virttest/qemu_monitor.py
  def cmd..
  .....
    r = self._get_response(q_id, timeout)
    if r is None:
                raise MonitorProtocolError("Received no response to QMP "
                                           "command '%s', or received a "
                                           "response with an incorrect id"
                                           % cmd)
            if "return" in r:
                ret = r["return"]
                if ret:
                    self._log_response(cmd, ret, debug)
                return ret
            if "error" in r:
                raise QMPCmdError(cmd, args, r["error"])

Output has been handled in def cmd, so no need to check the output here.
In this case, get response of query-named-block-nodes is just ok.

File: qemu/tests/blockdev_commit_query_named_block_nodes.py (hunk 6 above)
Comment (aliang123) [reply qemu/tests/blockdev_commit_query_named_block_nodes.py:24]: Will update test case.

File: qemu/tests/blockdev_commit_query_named_block_nodes.py (hunk 7 above)
Comment (aliang123) [reply qemu/tests/blockdev_commit_query_named_block_nodes.py]: Have updated.

**Comment by zhencliu** [comment]
LGTM

**Comment by ngu-niny** [comment]
Test PASS on ppc:

 (1/1) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.ppc64le.io-github-autotest-qemu.blockdev_commit_query_named_block_nodes: PASS (149.64 s)

Acked-by: Nini Gu <ngu@redhat.com>

**Pull request merged**